apple/swift-docc
# Add a new link resolver that finds documentation in a hierarchy of path components

Bug/issue #, if applicable: rdar://78518537 

## Summary

This adds an alternative link resolution implementation that searches a hierarchy of path components with disambiguation information to find documentation. This new resolver is off-by-default but can be enabeld by via a user default or environmental variable. 

This new implementation is meant to eventually replace the current documentation cache based implementation but only after we're confident in the quality. Since the documentation cache based link resolver implementation is meant to be removed in the future, I didn't see a value in defining a common interface for both implementations. 

Because both implementation exist side-by-side almost no code is removed in this PR. Breaking down the 21.7K added lines of code; 18.7K are from new symbol graph file test data and the remaining 3K LOC is half new implementation and half new unit tests.

This new implementation was originally prototyped here https://github.com/apple/swift-docc/pull/251. The user facing differences it brings is more flexibility with disambiguation for symbols and that more actionable information in diagnostics for unresolved links. 

## Dependencies

n/a

## Testing

Without any additional configuration DocC should continue to behave exactly the same.

To test the new link resolver implementation:
- Set either the `DocCUseHierarchyBasedLinkResolver` user default or the `DOCC_USE_HIERARCHY_BASED_LINK_RESOLVER` environmental variable to a true boolean value.
- Run any `docc convert` command. 

The expected behaviors are:
- All links that resolve in the default DocC configuration should continue to resolve
- Some links that didn't resolve in the default configuration may resolve successfully in the new implementation (for example if they are overly disambiguated).
- Links that fail to resolve (especially links to symbols) should have more actionable diagnostics.

You can also set the `DOCC_REPORT_LINK_RESOLUTION_MISMATCHES`  environmental variable to a true boolean value to enable automatic reporting of mismatched behaviors between the two implementations. Opting in to automatic reporting means that portions of both systems will run side-by-side performing redundant work so it comes with some performance cost.

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.
